dockerversion/useragent.go

```diff
@@ -20,5 +24,50 @@ func DockerUserAgent() string {
 	httpVersion = append(httpVersion, useragent.VersionInfo{Name: "os", Version: runtime.GOOS})
 	httpVersion = append(httpVersion, useragent.VersionInfo{Name: "arch", Version: runtime.GOARCH})
 
-	return useragent.AppendVersions("", httpVersion...)
+	dockerUA := useragent.AppendVersions("", httpVersion...)
+	if len(upstreamUA) > 0 {
+		ret := insertUpstreaUserAgent(upstreamUA, dockerUA)
+		return ret
+	}
+	return dockerUA
+}
+
+// GetUserAgentFromContext returns the previously saved user-agent context stored in ctx, if one exists
+func GetUserAgentFromContext(ctx context.Context) string {
+	var upstreamUA string
+	if ctx != nil {
+		var ki interface{} = ctx.Value(httputils.UAStringKey)
+		if ki != nil {
+			upstreamUA = ctx.Value(httputils.UAStringKey).(string)
+		}
+	}
+	return upstreamUA
+}
+
+// escapeStr returns s with every rune in charsToEscape escaped by a backslash
+func escapeStr(s string, charsToEscape string) string {
+	var ret string
+	for _, currRune := range s {
+		appended := false
+		for _, escapeableRune := range charsToEscape {
+			if currRune == escapeableRune {
+				ret += "\\" + string(currRune)
+				appended = true
+				break
+			}
+		}
+		if !appended {
+			ret += string(currRune)
+		}
+	}
+	return ret
+}
+
+// insertUpstreaUserAgent adds the upstream client useragent to create a user-agent
+// string of the form:
+//   $dockerUA UPSTREAM-CLIENT($upstreamUA)
+func insertUpstreaUserAgent(upstreamUA string, dockerUA string) string {
+	charsToEscape := "();\\" //["\\", ";", "(", ")"]string
+	upstreamUAEscaped := escapeStr(upstreamUA, charsToEscape)
+	return fmt.Sprintf("%s UPSTREAM-CLIENT(%s)", dockerUA, upstreamUAEscaped)
```

Comment:
I know this is bikeshedding, but why all caps? It feels a little out of place in a User-Agent header. I would have expected something like `UpstreamClient`.

I'm also wondering if there should be a space before the open paren. The sample User-Agent strings I looked at all have spaces before a parenthetical. But I'll defer to your and @stevvooe's research and expertise on this.

Anyway, this is cosmetic, so don't let my opinion derail things.